### PR TITLE
LruCache optimization

### PR DESCRIPTION
**Pull request by pizzi80**
I'm trying to realign my fork with official 4.x version

### CODE REVIEW

**Comment by pizzi80** [comment]
WAIT TO MERGE, IT's not completed, 
I've found some `ConcurrentModificationException`

File: src/main/java/org/omnifaces/util/cache/LruCache.java [hunk 1]
```diff
@@ -202,56 +194,72 @@ public boolean containsValue(Object value) {
 
     @Override
     public Set<K> keySet() {
-        return new ReadOnlyCollection<>(entries::keySet);
+        return unmodifiableSet(entries.keySet());
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/cache/LruCache.java:197]: Please revert. This is not concurrent.

File: src/main/java/org/omnifaces/util/cache/LruCache.java [hunk 2]
```diff
@@ -202,56 +194,72 @@ public boolean containsValue(Object value) {
...
     }
 
     @Override
     public Collection<V> values() {
-        return new ReadOnlyCollection<>(entries::values);
+        return unmodifiableCollection(entries.values());
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/cache/LruCache.java:202]: Please revert. This is not concurrent.

File: src/main/java/org/omnifaces/util/cache/LruCache.java [hunk 3]
```diff
@@ -202,56 +194,72 @@ public boolean containsValue(Object value) {
...
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public Set<Entry<K, V>> entrySet() {
-        return new ReadOnlyCollection<>(entries::entrySet);
+        return unmodifiableSet((Set<? extends Entry<K, V>>) entries.keySet());
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/cache/LruCache.java:208]: Please revert. This is not concurrent.

File: src/main/java/org/omnifaces/util/cache/LruCache.java [hunk 4]
```diff
@@ -202,56 +194,72 @@ public boolean containsValue(Object value) {
...
+        lock.lock();
 
-        @Override
-        public E next() {
-            return iterator.next();
+        try {
+            return task.get();
+        }
+        finally {
+            lock.unlock();
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/cache/LruCache.java:262]: Not sure these helpers are "optimization".

**Comment by BalusC** [comment]
Sorry, effort is appreciated, but I can't accept this. This all is changing the logic/intent.

**Comment by pizzi80** [comment]
I've tested the `LruCache` and both versions suffer from `ConcurrenModificationException`

during multiple threads insert and collections traversal,
I'm committing a test to demonstrate it ... and also a fix

**Pull request closed without merging**